# Homework

## Set up the environment

I used Github Codespaces and runned the following commands:

Create the environment
```
conda create -n ml-zoomcamp python=3.11
```

Activate it:

```
conda activate ml-zoomcamp
```

Installing libraries

```
conda install numpy pandas scikit-learn seaborn jupyter
```

# Q1. Pandas version

What version of Pandas did you install?

You can get the version information using the __version__ field:

In [37]:
pd.__version__

'2.3.2'

## Getting the data

In [86]:
import pandas as pd
import numpy as np

In [53]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

# Q2. Records count
How many records are in the dataset?

- [ ] 4704
- [ ] 8704
- [X] **9704 Correct Answer!**
- [ ] 17704

In [54]:
df.shape[0] # To get rows / records

9704

# Q3. Fuel types
How many fuel types are presented in the dataset?

- [ ] 1
- [x] **2  Correct Answer!**
- [ ] 3
- [ ] 4

In [55]:
# List of columns  
df.columns

Index(['engine_displacement', 'num_cylinders', 'horsepower', 'vehicle_weight',
       'acceleration', 'model_year', 'origin', 'fuel_type', 'drivetrain',
       'num_doors', 'fuel_efficiency_mpg'],
      dtype='object')

In [56]:
# nunique() Count number of distinct elements
df['fuel_type'].nunique() 

2

In [57]:
# unique() find the unique elements 
df['fuel_type'].unique() 

array(['Gasoline', 'Diesel'], dtype=object)

In [58]:
df.groupby('fuel_type').count()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
fuel_type                                                                   
Diesel                    4806           4601        4461            4806   
Gasoline                  4898           4621        4535            4898   

           acceleration  model_year  origin  drivetrain  num_doors  \
fuel_type                                                            
Diesel             4350        4806    4806        4806       4550   
Gasoline           4424        4898    4898        4898       4652   

           fuel_efficiency_mpg  
fuel_type                       
Diesel                    4806  
Gasoline                  4898

# Q4. Missing values
How many columns in the dataset have missing values?

- [ ] 0
- [ ] 1
- [ ] 2
- [ ] 3
- [X] **4 Correct Answer!**

In [59]:
# Sum of null values in each columns
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

# Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

- [ ] 13.75
- [x] 23.75 **Correct Answer!**
- [ ] 33.75
- [ ] 43.75

## Solution 1

In [60]:
df_asia=df[(df['origin'] == 'Asia')]
df_asia

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0         NaN     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12            15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14            14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20            18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21             NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   
...            ...         ...    ...       ...                ...        ...   
9688          15.5        2018   Asia    Diesel    All-wheel drive       -1.0   
9692          18.0        2016   Asia  Gasoline  Front-wheel drive        1.0   
9693          15.6        2012   Asia    Diesel    All-wheel drive        0.0   
9698          13.2        2005   Asia  Gasoline  Front-wheel drive       -2.0   
9703          14.7        2005   Asia    Diesel    All-wheel drive       -1.0   

      fuel_efficiency_mpg  
8               16.823554  
12              16.175820  
14              11.871091  
20              18.889083  
21              16.077730  
...                   ...  
9688            11.054830  
9692            11.711653  
9693            17.202782  
9698            13.933716  
9703            14.884467  

[3247 rows x 11 columns]

In [61]:
df_asia.fuel_efficiency_mpg.max()

23.759122836520497

## Solution 2

In [62]:
df[df['origin']== "Asia"]["fuel_efficiency_mpg"].max()

23.759122836520497

# Q6. Median value of horsepower
Find the median value of the horsepower column in the dataset.
Next, calculate the most frequent value of the same horsepower column.
Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
Now, calculate the median value of horsepower once again.
Has it changed?

- [X]  **Yes, it increased(Correct Answer!)**
- [ ] Yes, it decreased 
- [ ] No 

In [63]:
# Median of horspower column
horsepower_median=df.horsepower.median()
horsepower_median

149.0

In [69]:
# Mode of horspower column (most frequent value)
horsepower_mode=df.horsepower.mode()[0]
horsepower_mode

np.float64(152.0)

In [75]:
# Missing values in the horsepower collumn were filled using the most frequent value (horsepower_mode)
df['horsepower']=df['horsepower'].fillna(horsepower_mode)

In [78]:
# Sum of null values in each columns
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower               0
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [79]:
# Median of horspower column after null values were filled with the most frequent value (horsepower_mode)
horsepower_median_2=df.horsepower.median()
horsepower_median_2

152.0

# Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

Note: You just implemented linear regression. We'll talk about it in the next lesson.

- [ ] 0.051
- [ ] 0.51
- [ ] 5.1
- [ ] 51

In [85]:
# Select all cars from Asia and select only 'vehicle_weight','model_year'
cars_asia=df[(df['origin'] == 'Asia')][['vehicle_weight','model_year']]
cars_asia

vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]

In [87]:
x=cars_asia.head(7).to_numpy()
x

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])